src/universe.rs

```diff
@@ -94,24 +108,96 @@ impl SymbolStore {
         self.name_by_sym.push(None);
         sym
     }
+}
 
+impl Symbols for SymbolStore {
     /// Get the name of a symbol, if it has one.
-    pub fn get_symbol_name(&self, sym: Sym) -> Option<&str> {
+    fn get_symbol_name(&self, sym: Sym) -> Option<&str> {
         self.name_by_sym.get(sym.ord()).and_then(|n| n.as_deref())
     }
 
     /// Returns the number of symbols that have been allocated in this universe.
-    pub fn num_symbols(&self) -> usize {
+    fn num_symbols(&self) -> usize {
         self.name_by_sym.len()
     }
 }
 
+impl<'a> SymbolStorage for SymbolStore {
+    /// Return the symbol associated with the name, or allocate a fresh ID and associate it with the
+    /// given name.
+    fn get_or_insert_named(&mut self, name: &str) -> Sym {
+        self.sym_by_name.get(name).copied().unwrap_or_else(|| {
+            let sym = Sym::from_ord(self.name_by_sym.len());
+            let shared_name: Arc<str> = name.into();
+            self.name_by_sym.push(Some(shared_name.clone()));
+            self.sym_by_name.insert(shared_name, sym);
+            sym
+        })
+    }
+}
+
 impl Default for SymbolStore {
     fn default() -> Self {
         Self::new()
     }
 }
 
+/// Stores and allocates unique symbols on top of a symbol store, without modifying it.
+#[derive(Debug)]
+pub struct SymbolOverlay<'a> {
+    symbols: &'a SymbolStore,
+    overlay: SymbolStore,
+}
+
+impl<'a> SymbolOverlay<'a> {
+    pub fn new(symbols: &'a SymbolStore) -> Self {
+        Self {
+            symbols,
+            overlay: Default::default(),
+        }
+    }
+}
+
+impl SymbolStorage for SymbolOverlay<'_> {
+    /// Return the symbol associated with the name, or allocate a fresh ID and associate it with the
+    /// given name.
+    fn get_or_insert_named(&mut self, name: &str) -> Sym {
+        self.symbols
+            .sym_by_name
+            .get(name)
+            .copied()
+            .unwrap_or_else(|| {
+                self.overlay
+                    .sym_by_name
+                    .get(name)
+                    .copied()
+                    .unwrap_or_else(|| {
+                        let sym = Sym::from_ord(self.num_symbols());
+                        let shared_name: Arc<str> = name.into();
+                        self.overlay.name_by_sym.push(Some(shared_name.clone()));
+                        self.overlay.sym_by_name.insert(shared_name, sym);
+                        sym
+                    })
```

Comment:
Couldn't this just be `self.overlay.get_or_insert_named(name)` with an offset applied to the resulting `Sym` index?

Reply:
It could, but this usage of SymbolStore goes against its contract as soon as the indices are out of sync. I prefer to make it clear here that we're doing something special by writing code specific to this case.
In other words, the algorithm may be duplicated, but the purpose is different so it should stay separate.
I feel even reusing SymbolStore for the overlay is questionable, unless an offset would be built into SymbolStore.

Reply:
> this usage of SymbolStore goes against its contract as soon as the indices are out of sync

I wouldn't say that. If you shift its symbols up by some offset here, but then shift them back whenever you interact with the store again (like you do in `get_symbol_name`), then from the point of view of the overlay store, it's as if you had used the symbol directly.

Reply:
From the point of vier of the overlay, yes. From the point of view of the store, the store is inconsistent. I think this is borderline `unsafe`. I wanted to make anyone looking at the code know that this usage is not normal - it's a hack that relies on implementation details.
Initially I got caught in this myself, using a regular call and thinking the abstraction will take care of things, which cost me some debugging and feeling like I fell into my own trap.
Should I replace this block with the call regardless?

Reply:
Yes, I think it would still be nicer to use the existing function, and then shift the symbol ordinal together with a clear warning. To me, that makes the "funky stuff" way more clear than duplicating the code. In fact, on the first read, I actually missed that `Sym`s were implicitly shifted here, since the difference is really small and indirect (`self.num_symbols()` vs `self.overlay.num_symbols()` used for initializing the `Sym`, rather than a more obvious `Sym::from_ord(sym.ord() + self.symbols.len())` which directly shows that there's an offset involved).

As a bonus, the offsetting logic here then also aligns with the subtraction down in `get_symbol_name`.

Reply:
I retract my statements about unsafe. I read this conversation once again and I saw what you're proposing isn't actually breaking the contract of SymbolStorage. I'm having some deficiency thinking about the big picture lately.

This part is done as you suggested.